### ISSUE TITLE
Validator of operator

### ISSUE THREAD

**Issue by hqnghi88** [gama-platform/gama#2474]
**Is your request related to a problem? Please describe.**

Sorry if it has somewhere the solution.
Related with  #2473 , i find out that most of valiations in operators will throw RuntimeException
throw GamaRuntimeException.error(type + " has less than " + nbClasses + " colors", null);

**Describe the solution you'd like**

As in any Statement we have the Validator
	public static class CreateValidator implements IDescriptionValidator<StatementDescription> 
to assert a syntaxic validation.

**Describe alternatives you've considered**

I did not found a way to throw compilation error in validation phase. I found this but it require Eobject, in case of operator , there is no way to get syntac element?
new GamlCompilationError(message, IGamlIssue.WRONG_VALUE, null, false, false,new String[] {""});

**Comment by AlexisDrogoul** [comment]
Hi Nghi, before answering, can you please, when you post issues: (1) label them; (2) use the templates by removing the "example" lines. It makes things a bit difficult to read. 

So, adding custom validators to operators... Why not, but a lot of validation is already done, first at the syntactic level, before the statement validators are called, and by using the annotations you can put on the operators. They are maybe not complete, but at least it allows to have a common infrastructure for all the operators. What you are talking about is **semantic** validation, by the way. 

Do you have a specific scenario in mind ? #2473 is unrelated, because the annotations were not correct. GamaRuntimeExceptions are correctly named (they are emitted at **runtime**) and should not be used for validation. Actually they are not taken into account (hence the exception you got).

My concern here is to limit, as much as possible, the creation of objects during semantic validation. A lot of efforts has been put into this (for instance the validation of a project like MAELIA used to take up to 1 or 2 GB of memory, now reduced to ~300 MB), and I'm afraid having an army of small validators would make memory increase again, not to mention execution time. 

The few scenarios I can see in which dedicated validators could be necessary are those not already covered by the annotations (verification of the contents type of contents type of complex data types, for instance). So if you have some scenarios, I would be interested to see what can be done to cover them with new annotations -- or implement specific validators.

And to finish: the expression has no access to its underlying EMF description (for different reasons). The only structure able to deliver and collect compilation errors are the IDescriptions (i.e. statements). If you look at the GamlExpressionCompiler, you will see that it has a context which is an instance of IDescription. So if a validator was to be introduced, it would have (1) to be attached through annotions to the OperatorProto; (2) instantiated only once if possible (so a stateless validator); (3) fed with both the IDescription context and the various IExpression arguments; (4) run at the end of GamlExpressionFactory.createOperator().

**Comment by hqnghi88** [comment]
Dear Alexis,
Sorry for the templates, i saw a list of new templates. I will take time for labelling too.
In fact, i dont have a specific scenario because i did not developed as much operators as need. I also realize the unrelating with 2473 after your fix. I just tried to emit an error about semantic and did not found a way at operator level.
Concerning the memory for MAELIA, i think we have the annotation in gaml , i remember one for do not emit warning or sth like that ( @nowarning??) We can have other like @noOperatorCheck....
Because such big models are not popular. 
I agee with the scenarios that verify the contents type or whatsover. The R user should appreciate this enhance, as the operator 's validator can support the validation of external script (Rscript, and may be Python, someday).
And yes, i have take alook in that tree. But if once we accept the proposed solution, the error mark can be placed at beginning of the model? (yes it lost semantic meaning but it is real time validation).

**Comment by AlexisDrogoul** [comment]
All right. Let us experiment with this. I have created all the infrastructure necessary to add individual validators to operators. It follows the same scheme as the validator of statements (actually, I fusioned the two notions into IValidator, adding the possibility for IDescriptionValidators to veto the compilation by overriding another method than the simple "validate"). 

The new IOperatorValidator interface is fully commented and shouldn't be too complicated to implement. Here is a short and stupid example of use : 

```java
	public static class MultiplyValidator implements IOperatorValidator {

		@Override
		public boolean validate(final IDescription description, final EObject emfContext,
				final IExpression... arguments) {
			description.info("Hello, I am the multiply validator", GENERAL, emfContext);
			return true;
		}

	}

	@validator (MultiplyValidator.class)
	@operator (
			value = IKeyword.MULTIPLY,
			can_be_const = true,
			category = { IOperatorCategory.ARITHMETIC },
			concept = { IConcept.MATH, IConcept.ARITHMETIC })
	@doc (
			value = "Returns the product of the two operands.",
			masterDoc = true,
			usages = @usage (
					value = "if both operands are numbers (float or int), performs a normal arithmetic product and returns a float if one of them is a float.",
					examples = @example (
							value = "1 * 1",
							equals = "1")),
			see = { IKeyword.PLUS, IKeyword.MINUS, IKeyword.DIVIDE })
	public static Integer opTimes(final Integer a, final Integer b) {
		return a * b;
	}
```

Which gives: 

<img width="106" alt="expression" src="https://user-images.githubusercontent.com/579256/43180228-1b17278a-9001-11e8-95dd-c3858dcad674.png">
<img width="106" alt="info" src="https://user-images.githubusercontent.com/579256/43180229-1b45f704-9001-11e8-8be1-df1b07a61c6f.png">

Have fun but don't break everything, please !

**Comment by AlexisDrogoul** [comment]
Another example related to #2473: 

```java
	public static class BrewerValidator implements IOperatorValidator {

		@Override
		public boolean validate(final IDescription context, final EObject emfContext, final IExpression... arguments) {
			if (arguments[0].isConst()) {
				final Object palette = arguments[0].getConstValue();
				if (palette instanceof String) {
					final String p = (String) palette;
					final ColorBrewer brewer = ColorBrewer.instance();
					if (!brewer.hasPalette(p)) {
						context.error("Palette " + p + " does not exist. Available palette names are: "
								+ Arrays.toString(brewer.getPaletteNames()), UNKNOWN_ARGUMENT, emfContext);
						return false;
					}
				}
			}
			return true;
		}
```

I will commit these soon (incl. a convenience method getConstValue() that catches RuntimeExceptions).

**Comment by AlexisDrogoul** [comment]
Hi, I will not put such pragma as individual validators may be very important in some cases, and although because, now that they exist, I will progressively use them for the whole validation of operators (i.e. have a "base" validator instance of IValidator for the verification of types and a CompoundOperatorValidator that can incorporate this one as well as others, including the ones defined by developers in the annotation).
